### PR TITLE
balancer/randomsubsetting: Extend the unit tests in the randomsubsetting package.

### PR DESCRIPTION
**Pull request by marek-szews**
Add more unit tests to increase test coverage of 'randomsubsetting' package.

RELEASE NOTES:

- balancer/randomsubsetting: Implementation of additional UT.

### CODE REVIEW

File: balancer/randomsubsetting/randomsubsetting_ext_test.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	"fmt"
+	"testing"
+
+	xxhash "github.com/cespare/xxhash/v2"
+	"google.golang.org/grpc/resolver"
+
+	_ "google.golang.org/grpc/balancer/roundrobin" // For round_robin LB policy in tests
+)
+
+func (s) TestSubsettingEndpointsDomain(t *testing.T) {
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_ext_test.go]: This test is similar to the existing test `TestCalculateSubset_Simple`. 
You can remove this and add more test cases in `TestCalculateSubset_Simple` if you think of any.

File: balancer/randomsubsetting/randomsubsetting_ext_test.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+			subsetOfEndpoints := b.calculateSubset(tc.endpoints)
+			got := len(subsetOfEndpoints)
+			if fmt.Sprint(got) != fmt.Sprint(tc.want) {
+				t.Fatalf("subset size=%v; endpoints(%v) = %v; want %v", tc.want, tc.endpoints, got, tc.want)
+			}
+		})
+	}
+}
+
+func (s) TestUniformDistributionOfEndpoints(t *testing.T) {
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_ext_test.go]: Add a descriptive comment for this test.

Comment (Pranjali-2501) [reply balancer/randomsubsetting/randomsubsetting_ext_test.go]: Move this test to the existing test file `randomsubsetting_test.go`.

Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_ext_test.go]: Can you run the test ~10k times on your local to verify that the test is not flaky.

File: balancer/randomsubsetting/randomsubsetting_ext_test.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+	var (
+		subsetSize = 4
+		iteration  = 1600
+		diff       int
+	)
+
+	endpoints := makeEndpoints(16)
+	expected := iteration / len(endpoints) * subsetSize
+	diff = expected / 7 // allow ~14% difference
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_ext_test.go]: Can you explain the reasoning behind 14% error-rate.

Comment (Pranjali-2501) [reply balancer/randomsubsetting/randomsubsetting_ext_test.go]: Can you look at [this](https://github.com/grpc/grpc-go/blob/c7ec4d9ae3281bc57a8adce59b572e56965fb728/internal/testutils/roundrobin/roundrobin.go#L234-L253) and modify the test to calculate how many iterations are needed for statistical significance, instead of arbitrarily adding huge error rate to satisfy small range of iterations.

File: balancer/randomsubsetting/randomsubsetting_ext_test.go (hunk 1 above)
Comment (easwars) [reply balancer/randomsubsetting/randomsubsetting_ext_test.go]: +1

Yes, the only possible case that could be added to `TestCalculateSubset_Simple` is the case where the number of endpoints is strictly greater than the subset size.

File: balancer/randomsubsetting/randomsubsetting_ext_test.go (hunk 2 above)
Comment (easwars) [reply balancer/randomsubsetting/randomsubsetting_ext_test.go]: +1 to both the above comments. Please add comments about why the math is the way it is in the test.

Comment (easwars) [reply balancer/randomsubsetting/randomsubsetting_ext_test.go]: How often does this flake currently? We do not want to add tests that can flake. We want a test that is guaranteed to pass every time that it runs, and when it fails, it should really mean that there is a bug in the code.

File: balancer/randomsubsetting/randomsubsetting_ext_test.go (hunk 1 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_ext_test.go]: Indeed it is only one test case missing, i had renamed function and added to test set.

File: balancer/randomsubsetting/randomsubsetting_ext_test.go (hunk 2 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_ext_test.go]: Done. Appropriate commentary added.

File: balancer/randomsubsetting/randomsubsetting_ext_test.go (hunk 3 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_ext_test.go]: I carefully analyzed the test acceptance criteria and came to the conclusion that it was too naive. Finally I use the Chi-square goodness-of-fit test, standard statistical method used to validate whether a dataset follows a uniform distribution. It assesses if the observed frequencies in your data align with the expected frequencies of a theoretical distribution where every outcome has an equal chance of occurring.

File: balancer/randomsubsetting/randomsubsetting_ext_test.go (hunk 2 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_ext_test.go]: I have added two examples of tests for which the uniform distribution verification is negative.

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 4]
```diff
@@ -259,3 +260,247 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+		} else {
+			testPassed = true
+		}
+		reportBuilder.WriteString(fmt.Sprintf("Distribution is not uniform (χ²=%.2f > critical value=%.2f)", chi2, criticalValue))
+	} else {
+		if !dd.expectSuccess {
+			testPassed = false
+		}
+		reportBuilder.WriteString(fmt.Sprintf("Distribution is uniform (χ²=%.2f <= critical value=%.2f)", chi2, criticalValue))
+	}
```
[the remark this answers is not in the document]
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go]: Done

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 5]
```diff
@@ -259,3 +260,247 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+			b := &subsettingBalancer{
+				cfg:        &lbConfig{SubsetSize: tc.subsetSize},
+				hashSeed:   0,
+				hashDigest: xxhash.New(),
+			}
+			subsetOfEndpoints := b.calculateSubset(tc.endpoints)
+			got := len(subsetOfEndpoints)
+			if fmt.Sprint(got) != fmt.Sprint(tc.want) {
+				t.Fatalf("subset size=%v; endpoints(%v) = %v; want %v", tc.want, tc.endpoints, got, tc.want)
+			}
```
[the remark this answers is not in the document]
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go]: Done

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 6]
```diff
@@ -259,3 +260,247 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+					cfg:        &lbConfig{SubsetSize: uint32(L)},
+					hashSeed:   uint64(i ^ 3 + K*i + L),
+					hashDigest: xxhash.New(),
+				}
+				subsetOfEndpoints := lb.calculateSubset(endpoints)
+
+				for _, ep := range subsetOfEndpoints {
+					EndpointCount[ep.Addresses[0].Addr]++
+				}
+			})
```
[the remark this answers is not in the document]
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go]: Done

**Comment by Pranjali-2501** [comment]
Apologies for the delay, I'll take a look at it today.

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 7]
```diff
@@ -259,3 +260,241 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
 		t.Fatal("Timed out waiting for second child policy update")
 	}
 }
+
+func (s) TestSubsettingEndpointsSimply(t *testing.T) {
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_test.go:264]: Test `TestSubsettingEndpointsSimply` and existing test `TestCalculateSubset_Simple` are doing the same think.

As mentioned [here](https://github.com/grpc/grpc-go/pull/8781/changes/04b91933098be4bd7280f93ead5ee1e8e0a692a5#r2720381100) we can get rid of `TestSubsettingEndpointsSimply` completely and added a testcase in `TestCalculateSubset_Simple` in which the endpoints are strictly greater than subset size.

Something like this
```
func (s) TestCalculateSubset_Simple(t *testing.T) {
	tests := []struct {
		name       string
		endpoints  []resolver.Endpoint
		subsetSize uint32
		want       []resolver.Endpoint
	}{
		// existing code
		{
			name:       "SubsetSizeLessThanNumberOfEndpoints",
			endpoints:  makeEndpoints(15),
			subsetSize: 5,
			want:       makeEndpoints(5),
		},
	}

	// existing code
}
```

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 8]
```diff
@@ -259,3 +260,241 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+
+type DistributionData struct {
+	cardinality   int
+	mean          float64
+	sigma         float64
+	expectSuccess bool
+}
+
+func (s) TestUniformDistributionOfEndpoints(t *testing.T) {
+
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_test.go]: Remove this extra line.

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 9]
```diff
@@ -259,3 +260,241 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+			eps:        17,
+			subsetSize: 2,
+			iteration:  100,
+			positive:   false,
+		},
+	}
+
+	for _, tc := range testCases {
+		endpoints := makeEndpoints(tc.eps)
+		// From a set of N numbers, we randomly select K-times a subset of L numbers,
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_test.go]: Instead of keeping it as an inline comment inside the test loop, we should move it to the official Godoc for `TestUniformDistributionOfEndpoints` 

Something like this in the beginning:
```suggestion
// TestUniformDistributionOfEndpoints verifies that the random subsetting 
// policy achieves a uniform distribution across backends. From a set of N
// numbers, it randomly selects K-times a subset of L numbers, where L < N.
// Then it calculates how many times each number belonging to set N appears,
// compute the variance and standard deviation, and use a Chi-Square test to
// check whether the distribution is uniform.
```

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 10]
```diff
@@ -259,3 +260,241 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+		L := int(tc.subsetSize)
+		K := int(tc.iteration)
+
+		p := float64(L) / float64(N)         // Probability of x ∈ N being drawn p(x) = L / N
+		E := float64(K) * p                  // Expected Value (Mean) E(N) = K * p
+		variance := float64(K) * p * (1 - p) // Variance σ²(N) = K * p * (1 - p)
+		sigma := math.Sqrt(variance)         // Standard Deviation σ(N) = sqrt(σ²(N))
+
+		EndpointCount := make(map[string]int, N)
+		initEndpointCount(EndpointCount, endpoints)
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_test.go]: Since `initEndpointCount()` is a very simple helper used only once within `TestUniformDistributionOfEndpoints`, we could consider inlining this logic directly into the main test function.

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 11]
```diff
@@ -259,3 +260,241 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+func initEndpointCount(epCount map[string]int, eps []resolver.Endpoint) {
+	for _, ep := range eps {
+		epCount[ep.Addresses[0].Addr] = 0
+	}
+}
+
+// ChiSquareCriticalValue calculates the critical value for alpha (e.g., 0.05)
+// and degrees of freedom (df).
+func chiSquareCriticalValue(alpha float64, df float64) float64 {
+	// 1. Find the Z-score for the given alpha.
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_test.go]: Please remove `1.`/`2.` from the comments.

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 12]
```diff
@@ -259,3 +260,241 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+	// For alpha = 0.05 (95% confidence), Z is approx 1.64485
+	z := getZScore(1 - alpha)
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_test.go]: Same here.
`getZScore` is only used once within `chiSquareCriticalValue` and serves as a simple lookup table, we can inline this logic too.

**Comment by Pranjali-2501** [comment]
@arjan-bal and @easwars, could you please take a look at the statistical verification logic introduced in `TestUniformDistributionOfEndpoints`.

Thanks

**Comment by easwars** [comment]
> TestUniformDistributionOfEndpoints

We already have an implementation of this statistical verification logic here: https://github.com/grpc/grpc-go/blob/7136e99ee323c26984174eb3cec85c201fef9946/internal/testutils/roundrobin/roundrobin.go#L253

@arjan-bal knows this better than anyone else.

But we should try and see if this can reused here instead of reimplementing it.

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 7 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go:264]: Not exactly like that, function makeEndpoints(15) will produce the series of endpoints with a arithmetic order {endpoint_0, endpoint_1, ..., endpoint_14}. Same makeEndpoints(5) give us {endpoint_0, endpoint_1,.., endpoint_4}. But the result of LB.calculateSubset(eps) is unpredictable. One time we can received subset {endpoint_4, endpoint_5, endpoint_8, endpoint_11, endpoint_14} another time completely different. Validation of members of subset 99% will failed. I have left those two tests separated, because the first of them, check the content of set (boundary condition is meet so method always return the unchanged set - with a origin order) while the second test due to randomisation of content of subset, validate only the cardinality of the set (number of elements).

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 8 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go]: Done

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 9 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go]: Done and move to the beginning of function declaration.

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 10 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go]: Done

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 9 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go]: Done and move to the beginning of function declaration.

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 10 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go]: Done

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 11 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go]: Done

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 12 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go]: Done.

**Review by marek-szews — COMMENTED** [review]
Can you look at rework.

**Comment by easwars** [comment]
@marek-szews : Could you please merge changes from the `master` branch into this PR branch to get the CIs rolling again. Thanks.

**Comment by marek-szews** [comment]
I would like to ask you to accept my review.

**Comment by Pranjali-2501** [comment]
> I would like to ask you to accept my review.

Apologies for the delay, I'll review the changes today. Meanwhile could you please merge the master to get the CI running.

**Comment by easwars** [comment]
> I would like to ask you to accept my review.

We cannot review this PR until you merge master into your branch and have the tests pass.

**Comment by marek-szews** [comment]
Closing this PR because of too many conflicts. I've replaced it with a fresh [PR#9059](https://github.com/grpc/grpc-go/pull/9059).
